recipes/LibriSpeech/ASR/CTC/train_with_wav2vec_k2.py

```diff
@@ -0,0 +1,557 @@
+    import librispeech_prepare
+
+    # multi-gpu (ddp) save data preparation
+    run_on_main(
+        librispeech_prepare.prepare_librispeech,
+        kwargs={
+            "data_folder": hparams["data_folder"],
+            "tr_splits": hparams["train_splits"],
+            "dev_splits": hparams["dev_splits"],
+            "te_splits": hparams["test_splits"],
+            "save_folder": hparams["output_folder"],
+            "merge_lst": hparams["train_splits"],
+            "merge_name": "train.csv",
+            "skip_prep": hparams["skip_prep"],
+        },
+    )
+
+    run_on_main(
+        librispeech_prepare.download_librispeech_vocab_text,
+        kwargs={"destination": hparams["data_folder"]},
+    )
+
+    # here we create the datasets objects as well as tokenization and encoding
+    train_data, valid_data, test_datasets = dataio_prepare(hparams)
+
+    # Create the lexicon.txt for k2
+    run_on_main(
+        sbk2.lexicon.prepare_char_lexicon,
+        kwargs={
+            "lang_dir": hparams["lang_dir"],
+            "vocab_files": [hparams["vocab_file"]],
+            "extra_csv_files": [hparams["output_folder"] + "/train.csv"],
+            "add_word_boundary": hparams["add_word_boundary"],
+        },
+    )
+
+    caching = (
+        {"cache": False}
+        if "caching" in hparams and hparams["caching"] is False
+        else {}
+    )
+
+    # Create the lang directory for k2
+    run_on_main(
+        sbk2.prepare_lang.prepare_lang,
+        kwargs={
+            "lang_dir": hparams["lang_dir"],
+            "sil_prob": hparams["sil_prob"],
+            **caching,
+        },
+    )
+
+    # OpenSLR ngram models
+    if (
+        hparams["G_arpa"] + ".gz"
+        in librispeech_prepare.OPEN_SLR_11_NGRAM_MODELs
+        and hparams["G_rescoring_arpa"] + ".gz"
+        in librispeech_prepare.OPEN_SLR_11_NGRAM_MODELs
+        and (
+            hparams["compose_HL_with_G"]
+            or hparams["decoding_method"] == "whole-lattice-rescoring"
+        )
+    ):
+        run_on_main(
+            librispeech_prepare.download_openslr_librispeech_lm,
+            kwargs={
+                "destination": hparams["lm_dir"],
+                "rescoring_lm": hparams["decoding_method"]
+                == "whole-lattice-rescoring",
+            },
+        )
+    # SB ngram models
+    elif (
+        "sb" in hparams["G_arpa"]
+        and "sb" in hparams["G_rescoring_arpa"]
+        and (
+            hparams["compose_HL_with_G"]
+            or hparams["decoding_method"] == "whole-lattice-rescoring"
+        )
+    ):
+        run_on_main(
+            librispeech_prepare.download_sb_librispeech_lm,
+            kwargs={
+                "destination": hparams["lm_dir"],
+                "rescoring_lm": hparams["decoding_method"]
+                == "whole-lattice-rescoring",
+            },
+        )
+    # Train your ngram models
+    else:
+        output_arpa = os.path.join(hparams["lm_dir"], hparams["G_arpa"])
```

Comment:
(@Adel-Moumen and @mravanelli point of view required)if this is training the LM -- we are breaking the standard way of doing things in SB. If it's LM training, it should be exported to the LM folder.

Reply:
Some clarification here, this is only dataprep for kenlm training.

Reply:
As pointed out by @pchampio, this recipe is not training any LM. They are using LM already trained/download by the community. Maybe we should make sure that the documentation shows that ?

Reply:
OK, I got it. If you do not provide an LM, then this recipe will prepare everything so that you can train yours and then will expect that you provide the trained LM. This is totally something that should be covered inside of LibriSpeech/LM and not here...

Reply:
At least, IMHO, I would prefer this data LM prep logic inside of LibriSpeech/LM and in the train.py recipe for wav2vec2 only cover the training of the model.